### PR TITLE
SimpLL: Added support for detecting differences in macros by

### PR DESCRIPTION
**Pull request by lenticularis39**
extracting them from the source code, getting a list of used
macros on the corresponding line and comparing them.

Note: this currently does not work in tests, because it can't find the original source file. Also due to a planned change in the communication interface between SimpLL and diffkemp it does not pass the differences to diffkabi yet (they can be seen in the debug output of SimpLL).

### CODE REVIEW

**Comment by lenticularis39** [comment]
The macro difference printing works in most cases, excluding some when SimpLL finds the wrong line in one of the files when looking for the macros at the location of the difference.

File: diffkemp/diffkabi.py [hunk 1]
```diff
@@ -195,9 +195,14 @@ def text_indent(text, width):
                     called_res.second.callstack.replace(dest_path + "/", ""),
                     indent))
                 output.write("\n\n")
-            output.write(text_indent("Diff:\n", indent))
-            output.write(text_indent(
-                called_res.diff, indent))
+            if called_res.diff.strip() == "" and \
+               called_res.macro_diff is not None:
```
Comment (viktormalik) [inline diffkemp/diffkabi.py]: Use parentheses around multiline if condition (this is specified by PEP I think).

File: diffkemp/semdiff/result.py [hunk 2]
```diff
@@ -37,11 +37,49 @@ def __init__(self, name, filename=None, callstack=None):
...
+            for stack in [self.first_stack, self.second_stack]:
+                for macro in stack:
+                    res = res + "  from macro " + macro.name + " in file " + \
+                        macro.file + " on line " + str(macro.line) + "\n"
+                res = res + "  used in the function on line " + \
+                    (str(self.first_line) if stack is self.first_stack else
+                     str(self.second_line)) + "\n"
+                res = res + "    " + stack[0].body + "\n"
+                if stack is self.first_stack:
+                    res = res + "\n"
```
Comment (viktormalik) [inline diffkemp/semdiff/result.py]: I think using the style "Some {} string".format(x) is better readable than using "Some " + x + " string".

File: diffkemp/diffkabi.py (hunk 1 above)
Comment (lenticularis39) [reply diffkemp/diffkabi.py]: Interestingly flake8 does not detect this, despite the fact this surely is specified by PEP.

File: diffkemp/simpll/Utils.h [hunk 3]
```diff
@@ -88,4 +89,30 @@ AttributeList cleanAttributeList(AttributeList AL);
 /// Get a non-const pointer to a call instruction in a function
 CallInst *findCallInst(const CallInst *Call, Function *Fun);
 
-#endif //DIFFKEMP_SIMPLL_UTILS_H
\ No newline at end of file
+/// Gets C source file from a DIScope and the module.
+std::string getSourceFilePath(DIScope *Scope);
+
+/// Returned macros with value and location.
+struct MacroElement {
```
Comment (viktormalik) [inline diffkemp/simpll/Utils.h]: This could go into the new file.

File: diffkemp/simpll/Utils.h [hunk 4]
```diff
@@ -88,4 +89,30 @@ AttributeList cleanAttributeList(AttributeList AL);
...
+    // The body is in DebugInfo, parentMacro is a key in the map, therefore
+    // both can be stored by reference.
+    StringRef body, parentMacro;
+    // This is the line in the C source code on which the macro is located.
+    int line;
+    std::string sourceFile;
+};
+
+/// Gets all macros used on the line in the form of a key to value map.
+std::unordered_map<std::string, MacroElement> getAllMacrosOnLine(
```
Comment (viktormalik) [inline diffkemp/simpll/Utils.h]: This could go into the new file.

File: diffkemp/simpll/Utils.h [hunk 5]
```diff
@@ -88,4 +89,30 @@ AttributeList cleanAttributeList(AttributeList AL);
...
+    StringRef line, StringMap<StringRef> macroMap);
+
+/// Gets all macros used on a certain DILocation in the form of a key to value
+/// map.
+std::unordered_map<std::string, MacroElement> getAllMacrosAtLocation(
```
Comment (viktormalik) [inline diffkemp/simpll/Utils.h]: This could go into the new file.

File: diffkemp/simpll/ModuleComparator.h [hunk 6]
```diff
@@ -23,6 +23,18 @@
 
 using namespace llvm;
 
+struct MacroDifference {
```
Comment (viktormalik) [inline diffkemp/simpll/ModuleComparator.h]: This could go into the new file.

File: diffkemp/simpll/DifferentialFunctionComparator.cpp [hunk 7]
```diff
@@ -489,3 +494,83 @@ int DifferentialFunctionComparator::cmpMemset(const CallInst *CL,
...
             STyL->getStructName() != STyR->getStructName();
 }
+
+/// Finds macro differences at the locations of the instructions L and R and
+/// adds them to the list in ModuleComparator.
+/// This is used when a difference is suspected to be in a macro in order to
+/// include that difference into ModuleComparator, and therefore avoid an
+/// empty diff.
+void DifferentialFunctionComparator::findMacroDifferences(
+        const Instruction *L, const Instruction *R) const {
```
Comment (viktormalik) [inline diffkemp/simpll/DifferentialFunctionComparator.cpp]: Maybe even part of this function could go into the new file.

**Comment by lenticularis39** [comment]
Unfortunately Clang does not preserve ends of lines in the debug information about macros (it merges them into one line), therefore the exact location where the macro in stack was used cannot be determined (the line on which the macro is defined is used instead).

**Comment by lenticularis39** [comment]
The output now looks like this in case of a difference in a macro deeper in the include tree (the original line where the first macro in the tree is included is shown, too):
```
__bitmap_parse:
  __get_user_asm differs:
  {{{
    Callstack (3.10.0-862.el7):
    __get_user (macro) at lib/bitmap.c:450
    __get_user_nocheck (macro) at ./arch/x86/include/asm/uaccess.h:474
    __get_user_size (macro) at ./arch/x86/include/asm/uaccess.h:404
    __get_user_asm (macro) at ./arch/x86/include/asm/uaccess.h:335

    Callstack (3.10.0-957.el7):
    __get_user (macro) at lib/bitmap.c:450
    __get_user_nocheck (macro) at ./arch/x86/include/asm/uaccess.h:498
    __get_user_size (macro) at ./arch/x86/include/asm/uaccess.h:422
    __get_user_asm (macro) at ./arch/x86/include/asm/uaccess.h:347

    Diff:
      asm volatile(ASM_STAC "\n" "1:	mov"itype" %2,%"rtype"1\n" "2: " ASM_CLAC "\n" ".section .fixup,\"ax\"\n" "3:	mov %3,%0\n" "	xor"itype" %"rtype"1,%"rtype"1\n" "	jmp 2b\n" ".previous\n" _ASM_EXTABLE(1b, 3b) : "=r" (err), ltype(x) : "m" (__m(addr)), "i" (errret), "0" (err))
    
      asm volatile("\n" "1:	mov"itype" %2,%"rtype"1\n" "2:\n" ".section .fixup,\"ax\"\n" "3:	mov %3,%0\n" "	xor"itype" %"rtype"1,%"rtype"1\n" "	jmp 2b\n" ".previous\n" _ASM_EXTABLE(1b, 3b) : "=r" (err), ltype(x) : "m" (__m(addr)), "i" (errret), "0" (err))
  }}}
```
and like this when the function itself is deeper in the call stack (this is appended to the macro stack):
```
kmem_cache_alloc:
  for_each_object differs:
  {{{
    Callstack (3.10.0-862.el7):
    slab_alloc at mm/slub.c:2494
    slab_alloc_node at mm/slub.c:2489
    __slab_alloc at mm/slub.c:2449
    ___slab_alloc at mm/slub.c:2395
    deactivate_slab at mm/slub.c:2296
    discard_slab at mm/slub.c:1939
    free_slab at mm/slub.c:1523
    rcu_free_slab at mm/slub.c:1515
    __free_slab at mm/slub.c:1494
    for_each_object (macro) at mm/slub.c:1460

    Callstack (3.10.0-957.el7):
    slab_alloc at mm/slub.c:2534
    slab_alloc_node at mm/slub.c:2529
    __slab_alloc at mm/slub.c:2489
    ___slab_alloc at mm/slub.c:2435
    deactivate_slab at mm/slub.c:2336
    discard_slab at mm/slub.c:1979
    free_slab at mm/slub.c:1563
    rcu_free_slab at mm/slub.c:1555
    __free_slab at mm/slub.c:1534
    for_each_object (macro) at mm/slub.c:1500

    Diff:
      for (__p = (__addr); __p < (__addr) + (__objects) * (__s)->size; __p += (__s)->size)
    
      for (__p = fixup_red_left(__s, __addr); __p < (__addr) + (__objects) * (__s)->size; __p += (__s)->size)
  }}}
```

**Pull request merged**